File: 0020-valid-parentheses/0020-valid-parentheses.ts
function isValid(s: string): boolean {
    const stack = [];
    const openChars = ["(", "{", "["];
    const closeChars = [")", "}", "]"];
    const matchingPairs = {
        ")": "(",
        "}": "{",
        "]": "[",
    };

    for (let i = 0; i < s.length; i++) {
        if (openChars.includes(s.charAt(i))) {
            stack.push(s.charAt(i));
        } else if (closeChars.includes(s.charAt(i))) {
            const expectedOpenChar = matchingPairs[s.charAt(i)];
            if (stack.pop() !== expectedOpenChar) {
                return false;
            }
        } else {
            return false;
        }
    }

    return stack.length === 0;
};

File: 0020-valid-parentheses/0020-valid-parentheses.go
func isValid(s string) bool {
    pairs := map[rune]rune{
		'(': ')',
		'{': '}',
		'[': ']',
	}

    stack := []rune{}

    for _, char := range s {
        if _, ok := pairs[char]; ok {
            stack = append(stack, char)
        } else if len(stack) == 0 || pairs[stack[len(stack)-1]] != char {
            return false
        } else {
            stack = stack[:len(stack) - 1]
        }
    }

    return len(stack) == 0
}